--- cairo-1.12.0/util/cairo-missing/cairo-missing.h.orig	2012-02-02 01:48:15.000000000 +0100
+++ cairo-1.12.0/util/cairo-missing/cairo-missing.h	2012-03-25 12:56:29.970335651 +0200
@@ -37,7 +37,7 @@
 #include <string.h>
 #include <sys/types.h>
 
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(__MINGW32__)
 #define WIN32_LEAN_AND_MEAN
 #include <windows.h>

--- cairo-1.14.4/boilerplate/cairo-boilerplate-win32.c.orig	2015-10-27 22:04:22.000000000 +0100
+++ cairo-1.14.4/boilerplate/cairo-boilerplate-win32.c	2015-10-30 17:16:38.317127499 +0100
@@ -26,6 +26,7 @@
 
 #include "cairo-boilerplate-private.h"
 
+#include <limits.h>
 #include <cairo-win32.h>
 
 static const cairo_user_data_key_t win32_closure_key;
--- cairo-1.14.12/src/cairo-compiler-private.h.orig	2017-12-04 23:24:46.000000000 +0100
+++ cairo-1.14.12/src/cairo-compiler-private.h	2018-09-27 21:39:54.869047194 +0200
@@ -195,7 +195,7 @@
 #define __attribute__(x)
 #endif
 
-#if (defined(__WIN32__) && !defined(__WINE__)) || defined(_MSC_VER)
+#if (defined(__WIN32__) && !defined(__WINE__) && !defined(__MINGW32__)) || defined(_MSC_VER)
 #define access _access
 #define fdopen _fdopen
 #define hypot _hypot
