--- cairo-1.18.0/src/cairo-compiler-private.h.orig	2023-09-25 17:24:13.379273100 +0200
+++ cairo-1.18.0/src/cairo-compiler-private.h	2023-09-25 17:24:48.662415288 +0200
@@ -149,7 +149,7 @@
 #define __attribute__(x)
 #endif
 
-#if (defined(__WIN32__) && !defined(__WINE__)) || defined(_MSC_VER)
+#if (defined(__WIN32__) && !defined(__WINE__) && !defined(__MINGW32__)) || defined(_MSC_VER)
 #define access _access
 #ifndef R_OK
 #define R_OK 4
