--- cairo-1.0.2/configure.in.orig	2005-10-04 02:43:01.000000000 +0200
+++ cairo-1.0.2/configure.in	2006-03-04 21:24:15.043739500 +0100
@@ -94,7 +94,7 @@
     use_xlib=yes], [
       AC_CHECK_HEADER(X11/extensions/Xrender.h, [
       AC_PATH_XTRA
-      XRENDER_LIBS="$X_LIBS -lXrender -lXext -lX11 $X_EXTRA_LIBS"
+      XRENDER_LIBS="$X_LIBS -lXrender -lX11 $X_EXTRA_LIBS"
       use_xlib=yes], [
 		use_xlib="no (requires Xrender http://freedesktop.org/Software/xlibs)"])])
   _CHECK_FUNCS_WITH_FLAGS(XrmFinalize, $XRENDER_CFLAGS, $XRENDER_LIBS)
