--- cairo-1.8.10/configure.ac.orig	2010-02-19 19:38:03.000000000 +0100
+++ cairo-1.8.10/configure.ac	2010-02-23 08:25:04.000000000 +0100
@@ -43,7 +43,7 @@
 		     if test "x$no_x" = xyes; then
 		       use_xlib="no (requires X development libraries)"
 		     else
-		       xlib_NONPKGCONFIG_LIBS="$X_PRE_LIBS $X_LIBS -lX11 $X_EXTRA_LIBS"
+		       xlib_NONPKGCONFIG_LIBS="$X_LIBS -lX11 $X_EXTRA_LIBS"
 		       xlib_NONPKGCONFIG_CFLAGS=$X_CFLAGS
 		     fi])
 ])
